config/internal/configsource/component.go

```diff
@@ -21,32 +21,69 @@ import (
+	// configuration during time that data is being retrieved from the source.
 	//
 	// The code managing the returned Session object must guarantee that the object is not used
-	// concurrently.
+	// concurrently and that a single ConfigSource only have one Session open at any time.
 	NewSession(ctx context.Context) (Session, error)
 }
 
-// Session is the interface used to inject configuration data from a ConfigSource. A Session
+// Session is the interface used to retrieve configuration data from a ConfigSource. A Session
 // object is created from a ConfigSource. The code using Session objects must guarantee that
 // methods of a single instance are not called concurrently.
 type Session interface {
-	// Apply goes to the configuration source, and according to the specified selector and
-	// parameters retrieves a configuration value that is injected into the configuration.
+	// Retrieve goes to the configuration source and retrieves the selected data which
+	// contains the value to be injected in the configuration and the corresponding Watcher that
+	// will be used to monitor for updates of the retrieved value. The retrieved value is selected
+	// according to the selector and the params passed in the call to Retrieve.
 	//
-	// The selector is a string that is required on all invocations, the params are optional.
-	Apply(ctx context.Context, selector string, params interface{}) (interface{}, error)
+	// The selector is a string that is required on all invocations, the params are optional. Each
+	// implementation handles the generic params according to their requirements.
+	Retrieve(ctx context.Context, selector string, params interface{}) (RetrievedConfig, error)
 
-	// Close signals that the object won't be used anymore to inject data into a configuration.
-	// Each Session object should use this call according to their needs: release resources,
+	// RetrieveEnd signals that the Session must not be used to retrieve any new values from the
+	// source, ie.: all values from this source were retrieved for the configuration. It should
+	// be used to release resources that are only needed to retrieve configuration data.
+	RetrieveEnd(ctx context.Context) error
+
+	// Close signals that the configuration it was used to retrieve values is no longer in use
+	// and the object should close and release any watchers that it may have created.
+	// This method must be called when the configuration session ends, either in case of success
+	// or error. Each Session object should use this call according to their needs: release resources,
 	// close communication channels, etc.
 	Close(ctx context.Context) error
 }
+
+// RetrievedConfig holds the result of a call to the Retrieve method of a Session object.
+type RetrievedConfig struct {
+	// Value is the retrieved data that will be injected on the configuration.
+	Value interface{}
+	// Watcher monitors for updates of the retrieved value on the configuration source.
+	Watcher Watcher
```

Comment:
I think because watcher needs to be stopped at some point. We discussed a couple more ideas. 

1. session can derive  a new cancelable context from the context it receives and return a channel + close function instead of a Watcher interface. Somethng like:

```go

func (s *Session) RetrieveEnd(ctx context.Context) error {
     ...
     watchCtx, closeWatcher := context.WithCancel(ctx)
     ch := startWatcher(watchCtx)
     return RetrieveResult{
          Value: value,
          WatchCh: ch,
          CloseWatcher: closeWatcher,
     }
}
```

We can also further simplify by not returning a close/cancel function but letting the collector just cancel the context it passed to Session when it wants to stop a watcher. It doesn't feel as clean because a context passed to Session would ideally be used to cancel the Retrieve function itself.

Reply:
and then we'll not need `RetrieveEnd()` method in Session I think.

Reply:
I will remove the `Watcher` and move the function to the struct. The `RetrieveEnd` is not directly related to the watchers. Let's go without the cancellation token for the time being and see how it looks in a few implementations.